corgi/tasks/prod_defs.py

```diff
@@ -67,6 +69,87 @@ def update_products() -> None:
+        return []
+    regex_query = Q()
+    for pattern in cpe_patterns:
+        regex = pattern.replace(".", "\\.").replace("*", ".*")
+        regex_query |= Q(cpe__regex=regex)
+    return list(
+        CollectorErrataProductVariant.objects.filter(live=True)
+        .filter(regex_query)
+        .values_list("cpe", flat=True)
+    )
+
+
+def _match_stream_version_to_cpe_version(
+    version_streams: QuerySet[ProductStream],
+    cpe_version: str,
+    cpe: str,
+    el_by_cpe: defaultdict[str, list],
+    cpes_by_stream: defaultdict[str, list],
+):
+    """Iterate product stream versions, looking for matches to the cpe_version passed in.
+    If a match is found, store it in cpes_by_streams dict along with any el suffixes."""
+    for version_stream in version_streams:
+        if version_stream.version.removesuffix(".z") == cpe_version:
+            if cpe in el_by_cpe:
+                for el in el_by_cpe[cpe]:
+                    cpes_by_stream[version_stream.name].append(cpe + "::" + el)
+            else:
+                cpes_by_stream[version_stream.name].append(cpe)
+
+
+def _split_cpe_version(cpe_part: str, versions_by_cpe: defaultdict[str, set]):
+    version_split = cpe_part.rsplit(":", 1)
+    cpe_name = version_split[0]
+    cpe_version = version_split[-1]
+    versions_by_cpe[cpe_name].add(cpe_version)
+
+
+def _split_cpe(cpe: str, versions_by_cpe: defaultdict[str, set], el_by_cpe: defaultdict[str, list]):
+    """Splits a cpe such as 'cpe:/a:redhat:openshift_ironic:4.13:el9' into
+    'cpe:/a:redhat:openshift_ironic', '4.13', and 'el9'
+    It builds 2 dictionaries, one with a set of versions grouped by cpe_name and another
+    with el suffixes grouped by 'cpe_name + version'.
+    We do this so that we can iterate versions looking for matches without el suffixes. Once
+    we find a match, we add back the suffixes"""
+    variant_split = cpe.rsplit("::")
+    if len(variant_split) == 2:
+        el_part = variant_split[1]
+        cpe_part = variant_split[0]
+        el_by_cpe[cpe_part].append(el_part)
+        _split_cpe_version(cpe_part, versions_by_cpe)
+    elif len(variant_split) == 1:
+        _split_cpe_version(cpe, versions_by_cpe)
+    else:
+        raise ValueError(f"More than one el qualifier '::' in {cpe}")
+
+
+def _match_and_save_stream_cpes(product_version: ProductVersion) -> None:
+    """Given a Product Version with Product Stream foreign keys and cpes_matching_patterns
+    This will match those cpes to the stream children using the stream versions, and save the cpes
+    to the cpes_matching_patterns attribute of the streams."""
+    versions_by_cpe: defaultdict[str, set] = defaultdict(set)
+    el_suffix_by_cpe: defaultdict[str, list] = defaultdict(list)
+    for cpe in product_version.cpes_matching_patterns:
+        _split_cpe(cpe, versions_by_cpe, el_suffix_by_cpe)
+
+    cpes_by_stream: defaultdict[str, list] = defaultdict(list)
+    version_streams = product_version.productstreams.all()
```

Comment:
```suggestion
    version_streams = product_version.productstreams.get_queryset()
```

What happens if a stream is deleted in prod-defs? I guess maybe it never happens. But if it did, we'd keep the CPE for that stream's Variants on both the stream (which should be deleted) and the version (which should stop referencing the Variant CPEs from that deleted stream).

Reply:
I disagree. We replace the cpes_matching_patterns each time we run parse_product_version, so we wouldn't keep that stream's CPES.

Reply:
Oops, I think you're right about the version. Sorry for the confusion, I've been rather tired lately. But don't streams that disappear in prod-defs hang around in Corgi?

I guess at the end of the update_products() task, we need to find all the products, versions, streams, and variants which weren't updated (or created) by the task. We could add each object's PK / name to a set, then iterate over all Product / ProductVersion / ProductStream / ProductVariant.objects.exclude(name__in=set_of_processed_names) if any exist.

Any results in that queryset mean we have an orphaned model in Corgi, which doesn't exist anymore in prod-defs. This would handle deleted streams as well as (I think) deleted Brew tag Variants. Then we could log a warning, delete the orphaned model, or something else.

Maybe that needs to be a separate ticket, but I think we need to add this logic ASAP given how orphaned Brew tag Variants in Corgi caused issues, and how in the future orphaned streams (or CPEs of orphaned Variants) might get published to the Customer Portal (or end up in a published manifest) by mistake.

Reply:
Deleting a ps_update_stream does happen, there is a proposed change now for ossm (product-definitions/-/merge_requests/2234/). I updated CORGI-726 with your good idea here.